### PR TITLE
feat!: Support changing time scale

### LINKED ISSUE #85: Allow to change physics time scale
Following the issue https://github.com/jcornaz/heron/issues/81, that feature would allow us to run the defined physics_systems only on specific States. However, the physics engine itself will continue to be called.

A common feature for games is the ability to pause it at any time. In order to achieve that, we would need to be able to manually choose in which States the physics step will run. 

Another option is exposing a resource for allowing the user to start/stop the physics from any system. For this last option, it could be cool to also allow the user to modify the step used and how often it's called (i.e running the game in slow-motion while in pause to hurry up the player!).

### PR DESCRIPTION
**Pull request by yaymalaga**
Fix #85

### CODE REVIEW

**Comment by yaymalaga** [comment]
Some comments:
- Maybe it's worth it to make time_scale private again. Otherwise we can't avoid users setting negative values. 
- Ideally, heron systems shouldn't run if the step is 0.0. As far as I can think of, there is not need to calculate velocity, update bevy's position or check invalid bodies if the emulation is not running.

When I started testing ideas for the second point, the code below (core/src/lib) was working for me. However, for same reason, it is blocking now the game:
```rust
.............
            .add_stage_before(CoreStage::PostUpdate, crate::stage::ROOT, {
                let mut schedule = Schedule::default();

                if let Some(steps_per_second) = self.steps_per_second {
                    schedule = schedule
                        .with_run_criteria(FixedTimestep::steps_per_second(steps_per_second).chain(should_run.system()))
                }

                schedule.with_stage(crate::stage::UPDATE, SystemStage::parallel())
            });
    }
}

fn should_run(In(should_run): In<ShouldRun>, physics_controller: Res<'_, PhysicsController>) -> ShouldRun {
    if physics_controller.time_scale == 0.0 {
        ShouldRun::No
    } else {
        should_run
    }
}
```

**Comment by jcornaz** [comment]
Thanks for the contribution :-)

It looks good. I'll have a deeper look soon.

**Review by jcornaz — APPROVED** [review]
Everything looks good to me. Thanks @yaymalaga!

I made some small changes:
* Renamed the resource to `PhysicsTime`. I know the name you used was my suggestion. But after reading your comment, I realized `PhysicsTime` would be a better name.
* Added a simple tests for the `pause/resume` synergy.
* Simplified the `resume` logic to ignore the `scale` and only consider the `previous_scale`
  * I think the most important is that `resume` restore what was before `pause` (hence the added test). If the user do: `time.pause() ; time.scale = 0.5 ; time.resume()`, then I think the `resume` should ignore the `time.scale = 0.5` and still go back to the scale that was set before calling `pause`.


What do you think @yaymalaga ?

**Comment by jcornaz** [comment]
> Maybe it's worth it to make time_scale private again (revert commit bed7622). Otherwise we can't avoid users setting negative values.

Yes, actually. That's a good point. Maybe replace it by a setter.

> Ideally, heron systems shouldn't run if the step is 0.0. As far as I can think of, there is not need to calculate velocity, update bevy's position or check invalid bodies if the emulation is not running.

Yes, we could/should disable the `step` and `post_step` stages (in the rapier crate). But I wouldn't disable the whole physics root stage (which includes the user systems and the update of the rapier world).

The reasons is that disabling a stage where user-systems are running may be surprising for the user. Especially if the users are inexperimented or in the rush of a game-jam. For example if the system that call `resume` runs as a physics system, then the game will never be resumed^^.

Building on that, if we allow user systems to run, then we have to update the rapier world too. Otherwise we may miss updates, since the systems that update the rapier world are executed only when mutation are detected (in transform, velocity, etc.)

File: rapier/src/pipeline.rs [hunk 1]
```diff
@@ -16,7 +16,7 @@ use crate::rapier::pipeline::{ChannelEventCollector, PhysicsPipeline};
 pub(crate) fn step(
     mut pipeline: ResMut<'_, PhysicsPipeline>,
     gravity: Res<'_, Gravity>,
-    integration_parameters: Res<'_, IntegrationParameters>,
+    mut integration_parameters: ResMut<'_, IntegrationParameters>,
```
Comment (yaymalaga) [inline rapier/src/pipeline.rs:19]: Unfortunately we can't do this, take a look to the comment below

File: rapier/src/pipeline.rs [hunk 2]
```diff
@@ -28,7 +28,6 @@ pub(crate) fn step(
     physics_controller: Res<'_, PhysicsTime>,
 ) {
     let gravity = Vec3::from(*gravity).into_rapier();
-    let mut integration_parameters = integration_parameters.to_owned();
     integration_parameters.dt *= physics_controller.scale;
```
Comment (yaymalaga) [inline rapier/src/pipeline.rs:31]: As you are modifying the same instance over and over, you would be multiplying the scale in each step call. That means that a scale of 0.5 would turn into a 0 dt after a while. Basically, "cloning" the element was to just mutate it locally and avoid this problem.

**Comment by yaymalaga** [comment]
> 
> 
> Everything looks good to me. Thanks @yaymalaga!
> 
> I made some small changes:
> 
>     * Renamed the resource to `PhysicsTime`. I know the name you used was my suggestion. But after reading your comment, I realized `PhysicsTime` would be a better name.
> 
>     * Added a simple tests for the `pause/resume` synergy.
> 
>     * Simplified the `resume` logic to ignore the `scale` and only consider the `previous_scale`
>       
>       * I think the most important is that `resume` restore what was before `pause` (hence the added test). If the user do: `time.pause() ; time.scale = 0.5 ; time.resume()`, then I think the `resume` should ignore the `time.scale = 0.5` and still go back to the scale that was set before calling `pause`.
> 
> 
> What do you think @yaymalaga ?

Yeah everything looks good, thanks!!

**Comment by yaymalaga** [comment]
> > Maybe it's worth it to make time_scale private again (revert commit [bed7622](https://github.com/jcornaz/heron/commit/bed7622876f65552cd7f31292323265f2064ddcd)). Otherwise we can't avoid users setting negative values.
> 
> Yes, actually. That's a good point. Maybe use getter/setter.

Alright! I removed it initially because it wasn't that elegant, but it will remove some possible issues for sure.

> > Ideally, heron systems shouldn't run if the step is 0.0. As far as I can think of, there is not need to calculate velocity, update bevy's position or check invalid bodies if the emulation is not running.
> 
> Yes, we could/should disable the `step` and `post_step` stages (in the rapier crate). But I wouldn't disable the whole physics root stage (which includes the user systems and the update of the rapier world).

That sounds fair! In fact, if you take a look to Rapier's step function, it has an asset for dt > 0, so it would be wise to just skip the system.

> I think that disabling a stage where user-systems are running may be surprising for the user. Especially if the users are inexperimented or in the rush of a game-jam. For example if the system that call `resume` runs as a physics system, then the game will never be resumed^^.
> 
> Building on that, if we allow user systems to run, then we have to update the rapier world too. Otherwise we may miss changes, as the systems that update the rapier world are executed only when mutation are detected (in transform, velocity, etc.)

Yeah, you are totally right, we should just stop the minimal systems needed to avoid possible breaks. We don't even know if users are making use of any Heron's stage.

File: rapier/src/pipeline.rs (hunk 2 above)
Comment (jcornaz) [reply rapier/src/pipeline.rs:31]: Ah yes indeed. Thanks for pointing out.

**Comment by yaymalaga** [comment]
I implemented what we discussed, so this should be now ready to go!

**Comment by jcornaz** [comment]
Thanks @yaymalaga :-)

I'll check that (and probably merge) tonight.

**Comment by yaymalaga** [comment]
If we go with the heron-step and heron-post-step run criteria, we should warn in the changelog that these stages won't run when scale is set to 0. We don't know if people are adding systems to these stages.

**Comment by jcornaz** [comment]
> We don't know if people are adding systems to them.

In theory they shouldn't. Heron doesn't export any public acces to these stages. Of course,  a user can still read heron source code and find the stage name. But that would be relying on internal details and not on public API.

**Review by jcornaz — APPROVED** [review]
All good to me.

I did a last fine-tuning of how `PhysicsTime` should react to negative scale.

* `PhysicsTime::new` and `PhysicsTime::set_scale` will panic if the scale is negative.
* `PhysicsTime` no longer implement `From` because implementation of `From` shouldn't fail.
* `PhysicsTime::new` and `PhysiscsTime::set_scale` accept a scale of `0.0`. I think it would be akward to force the user to use `pause` for the specific case of a `0.0` scale. However `resume` only works when the user called `pause` (not when he/she set to `0.0`).

**Comment by yaymalaga** [comment]
Awesome, looks good to me :)

**Pull request merged**